File: day2.py
from collections import *

def part1():
    ans = 0
    cap = {
        "red": 12,
        "green": 13,
        "blue": 14
    }
    while line := input():
        game, data = line.split(":")
        game = int(game.split()[1])
        flag = True
        for draw in data.split(";"):
            for color in draw.split(","):
                amt, clr = color.split()
                flag &= int(amt) <= cap[clr]
        if flag:
            ans += game
    print(ans)

def part2():
    ans = 0
    while line := input():
        game, data = line.split(":")
        floor = defaultdict(int)
        for draw in data.split(";"):
            for color in draw.split(","):
                amt, clr = color.split()
                floor[clr] = max(floor[clr], int(amt))
        ans += floor["red"] * floor["green"] * floor["blue"]
    print(ans)

part1()
part2()

File: day2.cpp
#include <bits/stdc++.h>
using namespace std;
#define INPUT(var) for(string var; getline(cin, var), var!="";)
#define FOR(i, s, e) for (int i=s; i<e; ++i)

void part1() {
	map<string, int> cap{
		{"red", 12},
		{"green", 13},
		{"blue", 14}
	};
	int ans = 0;
	INPUT(line) {
		stringstream ss(line);
		string token;
		ss >> token >> token;

		token.pop_back();
		int game = stoi(token);

		bool flag = 1;
		while (ss >> token) {
			int n = stoi(token);
			ss >> token;
			if (!isalpha(token.back()))
				token.pop_back();

			flag &= n <= cap[token];
		}
		if (flag) ans += game;
	}
	cout << ans << '\n';
}

void part2() {
	int ans = 0;
	INPUT(line) {
		stringstream ss(line);
		string token;
		ss >> token >> token;

		map<string, int> floor;
		while (ss >> token) {
			int n = stoi(token);
			ss >> token;
			if (!isalpha(token.back()))
				token.pop_back();

			floor[token] = max(floor[token], n);
		}
		ans += floor["red"] * floor["green"] * floor["blue"];
	}
	cout << ans << '\n';
}

int main() {
	part1();
	part2();
}